engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java

```diff
@@ -432,4 +496,66 @@ private DirectBuffer insertAt(
     }
     return Either.right(false);
   }
+
+  private DirectBuffer wrapNumberOfElementInstances(
+      final MutableDirectBuffer variableBuffer,
+      final DirectBuffer variableView,
+      final long numberOfElementInstances) {
+    variableWriter.wrap(variableBuffer, 0);
+
+    variableWriter.writeInteger(numberOfElementInstances);
+    final var length = variableWriter.getOffset();
+
+    variableView.wrap(variableBuffer, 0, length);
+    return variableView;
+  }
+
+  private void updateNumberOfCompletedElementInstances(final BpmnElementContext context) {
+    updateNumberOfElementInstances(
+        context,
+        NUMBER_OF_COMPLETED_ELEMENT_INSTANCES_VARIABLE,
+        numberOfCompletedElementInstancesVariableBuffer,
+        numberOfCompletedElementInstancesVariableView);
```

Comment:
❌  I don't think the `numberOfCompletedElementInstancesVariableView` is ever changed, so it doesn't keep track of this value yet. It is only initialized. We'll need some tests that show that the value is actually kept up to date with the reality.

To achieve it, we'll need to do something similar to what you already do in `updateNumberOfActiveElementInstances`, but then for the number of completed element instances. This is where things get more complicated.

The number of completed element instances is not tracked in `ElementInstance`, like the number of active children is. I see 2 solution directions:
1. read the current `numberOfCompletedElementInstances` local variable's value.
2. add the number of completed and terminated element instances as properties to `ElementInstance` and read them like you do the `numberOfActiveElementInstances` in `updateNumberOfActiveElementInstances`. 

I'd be curious if you see any other ways to solve this, but in any case I can already give you my thoughts on the solutions I suggest here. My initial feeling is that reading the local variable is bit error prone, because the user could alter this value leading to incorrect readings. 

Solution 2 sounds a bit better, because it means we get to keep track of the value in the state rather than through a user-modifiable variable, but there are other problems with it. Currently, `ElementInstance` only keeps track of the child count using `childCountProp`, we'll need to add something that represents the number of completed children. But the `ElementInstance` represents **any** element instance, not just multi-instance bodies. So we'd need to consider whether those new props should also be updated for **non-multi-instance** elements as well and  that might slow down processing. We'd need to verify this in some benchmarks, but let's postpone that for now and focus on getting it working first. We can optimize it later.

Lastly, we'll need to consider a migration path because existing Zeebe clusters will have multi-instance elements active that don't yet have these properties and did not yet keep track of it. This is true for both solution directions. Do we want to add them retroactively or simply ignore them for existing active multi-instances? I'd be happy to hear your thoughts on this.

Reply:
Hi @korthout , As you mentioned, I think the second solution is more reasonable. Maybe we only need to add the number of completed element instances as properties to `ElementInstance`.

As we can see from page [#224](https://www.omg.org/spec/BPMN/2.0/PDF#page224), the numberOfTerminatedInstances variable of multi-instances are defined in the spec.
> The sum of `numberOfTerminatedInstances`, `numberOfCompletedInstances`, and `numberOfActiveInstances` always sums up to `numberOfInstances`.

So we can easily get the value of the `numberOfTerminatedInstances` element.

Reply:
That's an interesting idea. 

The team has been increasingly more interested in adding data checks in the engine to make sure we don't end up in a corrupted state. This specific requirement may actually be an interesting property to check. Having all 4 variables available in state would allow us to add a consistency check for it. So that would be an argument for keeping track of both the `numberOfCompletedInstances` as well as the `numberOfTerminatedInstances `.

Reply:
Hi @korthout , Do I need to wait for the consistency check capability to be available before proceeding with follow-up work?

Reply:
Hi @Lzgabel, I'm back from holidays 👋 

>Do I need to wait for the consistency check capability

No, I guess we had a miscommunication here. I mentioned this check as an argument to keeping track of both the `numberOfCompletedInstances` as well as the `numberOfTerminatedInstances` for each `ElementInstance`. 

However, adding these properties is complicated like I described. Therefore, I highly recommend to remove the `numberOfCompletedInstances` and the `numberOfTerminatedInstances` from this PR, and only focus on adding the `numberOfInstances` and the `numberOfActiveInstances` variables. That would allow us to quicker get new functionality merged as well as focus the other PR on the difficulties involved with the other variables. I hope that makes sense 🙇 

As always, please let me know if I can help with anything or if you have any questions. By the way, your contributions for completion condition will soon be available in release 1.3.0 🎉